crates/fce-test-macro-impl/src/fce_test/module_generator/methods_generator.rs

```diff
@@ -0,0 +1,161 @@
+    let ret = generate_ret(&output_type);
+
+    let function_call = quote! {
+        use std::ops::DerefMut;
+
+        #convert_arguments
+
+        #set_result #function_call
+
+        #convert_result_to_output_type
+
+        #ret
+    };
+
+    Ok(function_call)
+}
+
+/// Generates type convertor to json because of AppService receives them in json.
+fn generate_arguments_converter<'a>(
+    args: impl ExactSizeIterator<Item = &'a str>,
+) -> TResult<TokenStream> {
+    let arg_idents: Vec<syn::Ident> = args.map(utils::new_ident).collect::<Result<_, _>>()?;
+
+    let args_converter =
+        quote! { let arguments = fluence_test::internal::json!([#(#arg_idents),*]); };
+
+    Ok(args_converter)
+}
+
+fn generate_function_call(module_name: &str, method_name: &str) -> TokenStream {
+    quote! { self.fce.as_ref().borrow_mut().call_with_module_name(#module_name, #method_name, arguments, <_>::default()).expect("call to FCE failed"); }
+}
+
+fn generate_set_result(output_type: &Option<&IType>) -> TokenStream {
+    match output_type {
+        Some(_) => quote! { let result = },
+        None => TokenStream::new(),
+    }
+}
+
+fn generate_convert_to_output(
+    output_type: &Option<&IType>,
+    records: &FCERecordTypes,
+) -> TResult<TokenStream> {
+    let result_stream = match output_type {
+        Some(ty) => {
+            let ty = utils::itype_to_tokens(ty, records)?;
+            quote! {
+                let result: #ty = serde_json::from_value(result).expect("the default deserializer shouldn't fail");
+            }
+        }
+        None => TokenStream::new(),
+    };
+
+    Ok(result_stream)
+}
+
+fn generate_ret(output_type: &Option<&IType>) -> TokenStream {
+    match output_type {
+        Some(_) => quote! { result },
+        None => TokenStream::new(),
+    }
+}
+
+fn generate_arguments<'a, 'r>(
+    arguments: impl ExactSizeIterator<Item = &'a IFunctionArg>,
+    records: &'r FCERecordTypes,
+) -> TResult<Vec<TokenStream>> {
+    arguments
+        .map(|argument| -> TResult<_> {
+            let arg_name = utils::new_ident(&argument.name)?;
+            let arg_type = utils::itype_to_tokens(&argument.ty, records)?;
+
+            let arg = quote! { #arg_name: #arg_type };
+            Ok(arg)
+        })
+        .collect::<TResult<Vec<_>>>()
+}
+
+fn generate_output_type(output_types: &[IType], records: &FCERecordTypes) -> TResult<TokenStream> {
+    let output_type = get_output_type(output_types);
+    match output_type {
+        None => Ok(TokenStream::new()),
+        Some(ty) => {
+            let output_type = utils::itype_to_tokens(&ty, records)?;
+            let output_type = quote! { -> #output_type };
+
+            Ok(output_type)
+        }
+    }
+}
+
+fn get_output_type(output_types: &[IType]) -> Option<&IType> {
+    match output_types.len() {
+        0 => None,
+        1 => Some(&output_types[0]),
+        _ => unimplemented!("function with more than 1 arguments aren't supported now"),
```

Comment:
Should it be `more than 1 outputs/return values`?

Reply:
Only after switching to a VM with multi-value